In [1]:
import tweepy
import pandas as pd
import numpy as np
import json
import requests

## GATHER

In [2]:
df = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
#to see all the data
df

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [5]:
# to get random observations
#we see that there are some missing values
df.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1426  697881462549430272                    NaN                  NaN   
122   869227993411051520                    NaN                  NaN   
2152  669680153564442624                    NaN                  NaN   
1754  678798276842360832                    NaN                  NaN   
2138  670040295598354432                    NaN                  NaN   
1244  711743778164514816                    NaN                  NaN   
611   797165961484890113           7.971238e+17           29166305.0   
1375  701889187134500865                    NaN                  NaN   
1626  684800227459624960                    NaN                  NaN   
2299  667065535570550784                    NaN                  NaN   

                      timestamp  \
1426  2016-02-11 20:34:41 +0000   
122   2017-05-29 16:24:37 +0000   
2152  2015-11-26 00:52:45 +0000   
1754  2015-12-21 04:44:55 +0000   
2138  2015-11-27 00:43:49 +0000   
1244  2016-03-21 02:38:34 +0000   
611   2016-11-11 19:55:50 +0000   
1375  2016-02-22 21:59:57 +0000   
1626  2016-01-06 18:14:31 +0000   
2299  2015-11-18 19:43:11 +0000   

                                                 source  \
1426  <a href="http://twitter.com/download/iphone" r...   
122   <a href="http://twitter.com/download/iphone" r...   
2152  <a href="http://twitter.com/download/iphone" r...   
1754  <a href="http://twitter.com/download/iphone" r...   
2138  <a href="http://twitter.com/download/iphone" r...   
1244  <a href="http://twitter.com/download/iphone" r...   
611   <a href="http://twitter.com/download/iphone" r...   
1375  <a href="http://twitter.com/download/iphone" r...   
1626  <a href="http://twitter.com/download/iphone" r...   
2299  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1426  This is Carter. He wakes up in the morning and...                  NaN   
122   This is Gizmo. His favorite thing is standing ...                  NaN   
2152  This is Shawwn. He's a Turkish Gangrene Robitu...                  NaN   
1754  This is Linda. She fucking hates trees. 7/10 h...                  NaN   
2138  😂😂😂 10/10 for the dog and the owner https://t....                  NaN   
1244  Meet Roosevelt. He's calculating the best case...                  NaN   
611   @JODYHiGHROLLER it may be an 11/10 but what do...                  NaN   
1375  This is Bilbo. He's not emotionally prepared t...                  NaN   
1626  Meet Theodore. He's dapper as hell. Probably o...                  NaN   
2299  Here we have a Hufflepuff. Loves vest. Eyes wi...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1426                       NaN                        NaN   
122                        NaN                        NaN   
2152                       NaN                        NaN   
1754                       NaN                        NaN   
2138                       NaN                        NaN   
1244                       NaN                        NaN   
611                        NaN                        NaN   
1375                       NaN                        NaN   
1626                       NaN                        NaN   
2299                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1426  https://twitter.com/dog_rates/status/697881462...                10   
122   https://twitter.com/dog_rates/status/869227993...                13   
2152  https://twitter.com/dog_rates/status/669680153...                 8   
1754  https://twitter.com/dog_rates/status/678798276...                 7   
2138  https://twitter.com/dog_rates/status/670040295...                10   
1244  https://twitter.com/dog_rates/status/711743778...                11   
611                                                 NaN                11   
1375  https://twit

In [5]:
# now to programmatically download image_predictions.tsv
image = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')
with open('image_prediction.tsv','wb') as file:
    file.write(image.content)

In [6]:
df_images = pd.read_csv('image_prediction.tsv',sep = '\t')

In [7]:
#to see the data we collected
df_images

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [196]:
df_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [20]:
#to get random data
df_images.sample(10)

tweet_id                                          jpg_url  \
83    667495797102141441  https://pbs.twimg.com/media/CUNr4-7UwAAg2lq.jpg   
1954  864197398364647424  https://pbs.twimg.com/media/C_4-8iPV0AA1Twg.jpg   
1619  802624713319034886  https://pbs.twimg.com/media/CsrjryzWgAAZY00.jpg   
1727  820446719150292993  https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
780   689977555533848577  https://pbs.twimg.com/media/CZNK7NpWwAEAqUh.jpg   
1223  744709971296780288  https://pbs.twimg.com/media/ClW9w7mWEAEFN1k.jpg   
1649  809084759137812480  https://pbs.twimg.com/media/CzpyM41UoAE1b2w.jpg   
907   700462010979500032  https://pbs.twimg.com/media/CbiKe7-W0AIVNNr.jpg   
505   675891555769696257  https://pbs.twimg.com/media/CWE_x33UwAEE3no.jpg   
645   681610798867845120  https://pbs.twimg.com/media/CXWRZBgWkAEHMea.jpg   

      img_num                 p1   p1_conf  p1_dog                  p2  \
83          1          Chihuahua  0.143957    True  Christmas_stocking   
1954        4   golden_retriever  0.945905    True  Labrador_retriever   
1619        1     cocker_spaniel  0.253442    True    golden_retriever   
1727        3   golden_retriever  0.938048    True              kuvasz   
780         1         cowboy_hat  0.291081   False  Labrador_retriever   
1223        1  Shetland_sheepdog  0.234431    True             Samoyed   
1649        1             vizsla  0.911412    True          bloodhound   
907         1            hamster  0.678651   False          Pomeranian   
505         1  Italian_greyhound  0.305637    True             whippet   
645         1         toy_poodle  0.821704    True    miniature_poodle   

       p2_conf  p2_dog                  p3   p3_conf  p3_dog  
83    0.118651   False            ski_mask  0.092482   False  
1954  0.021264    True     Tibetan_mastiff  0.020493    True  
1619  0.162850    True          otterhound  0.110921    True  
1727  0.025119    True  Labrador_retriever  0.022977    True  
780   0.179625    True            sombrero  0.121493   False  
1223  0.114876    True              collie  0.086614    True  
1649  0.017134    True  Labrador_retriever  0.011761    True  
907   0.110268    True              Angora  0.104139   False  
505   0.232057    True          Great_Dane  0.117806    True  
645   0.116042    True   Yorkshire_terrier  0.014847    True

In [197]:
# to get consumer_key, consumer_secret, access_token and access_secret from twitter for WeRateDogs dataset
consumer_key = 'CONSUMER KEY'
consumer_secret = 'CONSUMER SECRET'
access_token = 'ACCESS TOKEN'
access_secret = 'ACCESS SECRET'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

# Construct the API instance firstly to parse the result to json object, secondly to wait for rate limits and lastly to print
#a notification when tweepy is waiting for rate limits
api = tweepy.API(auth, 
                 parser = tweepy.parsers.JSONParser(), 
                 wait_on_rate_limit = True, 
                 wait_on_rate_limit_notify = True) 

In [38]:
tweet_json=open('tweet_json.txt','w')
#txt_file.write('tweet_id retweet_count favorite_count''\n')
for tweet_id in df.tweet_id:
    try:
        tweet_info=api.get_status(str(tweet_id),tweet_mode='extended')
        retweet_count=tweet_info.retweet_count
        print(tweet_id)
        favorite_count=tweet_info.favorite_count
        tweet_json.write(tweet_id+' '+str(retweet_count)+' '+str(favorite_count)+'\n')  
    except Exception as e:
        print('error')
tweet_json.close()

error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
erro

error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
error
erro

In [8]:
df_tweet_json = pd.DataFrame(columns=['tweet_id','retweet_count','favorite_count'])
with open('tweet_json.txt','r')as file:
    for line in file:
        tweet_id,retweet_count,favorite_count=line[:-1].split(' ')
        df_tweet_json = df_tweet_json.append({'tweet_id':tweet_id,
                                              'retweet_count':retweet_count,
                                              'favorite_count':favorite_count},
                                               ignore_index=True)

In [9]:
df_tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2186 entries, 0 to 2185
Data columns (total 3 columns):
tweet_id          2186 non-null object
retweet_count     2186 non-null object
favorite_count    2186 non-null object
dtypes: object(3)
memory usage: 51.3+ KB


In [10]:
df_tweet_json.sample(10)

tweet_id retweet_count favorite_count
279   837366284874571778          5960          22946
446   815990720817401858          1200           5525
1439  687704180304273409           950           2651
378   824325613288833024         11779          12946
474   813051746834595840          8449          23204
774   768609597686943744          1377           4557
590   794926597468000259          2682          11451
1343  695064344191721472           677           1759
813   711694788429553666         20370          35663
1744  674468880899788800          2255           6633

## ACCESS

In [11]:
df.retweeted_status_id

0                NaN
1                NaN
2                NaN
3                NaN
4                NaN
5                NaN
6                NaN
7                NaN
8                NaN
9                NaN
10               NaN
11               NaN
12               NaN
13               NaN
14               NaN
15               NaN
16               NaN
17               NaN
18               NaN
19      8.874740e+17
20               NaN
21               NaN
22               NaN
23               NaN
24               NaN
25               NaN
26               NaN
27               NaN
28               NaN
29               NaN
            ...     
2326             NaN
2327             NaN
2328             NaN
2329             NaN
2330             NaN
2331             NaN
2332             NaN
2333             NaN
2334             NaN
2335             NaN
2336             NaN
2337             NaN
2338             NaN
2339             NaN
2340             NaN
2341             NaN
2342         

In [12]:
df.name.value_counts()

None       745
a           55
Charlie     12
Cooper      11
Lucy        11
Oliver      11
Lola        10
Penny       10
Tucker      10
Bo           9
Winston      9
Sadie        8
the          8
Toby         7
Bailey       7
Buddy        7
Daisy        7
an           7
Milo         6
Stanley      6
Leo          6
Scout        6
Jax          6
Dave         6
Oscar        6
Rusty        6
Jack         6
Koda         6
Bella        6
Chester      5
          ... 
Kanu         1
Rueben       1
Gustav       1
Howie        1
Lillie       1
Zoe          1
Tedrick      1
Julio        1
Karl         1
Arnold       1
Stella       1
such         1
Marlee       1
Jareld       1
Dotsy        1
Odin         1
Pip          1
Stu          1
Ike          1
Bruno        1
Tug          1
Mutt         1
Andru        1
Hall         1
Tommy        1
Jeffrie      1
Beya         1
Tassy        1
Siba         1
Spanky       1
Name: name, Length: 957, dtype: int64

In [13]:
df.text

0       This is Phineas. He's a mystical boy. Only eve...
1       This is Tilly. She's just checking pup on you....
2       This is Archie. He is a rare Norwegian Pouncin...
3       This is Darla. She commenced a snooze mid meal...
4       This is Franklin. He would like you to stop ca...
5       Here we have a majestic great white breaching ...
6       Meet Jax. He enjoys ice cream so much he gets ...
7       When you watch your owner call another dog a g...
8       This is Zoey. She doesn't want to be one of th...
9       This is Cassie. She is a college pup. Studying...
10      This is Koda. He is a South Australian decksha...
11      This is Bruno. He is a service shark. Only get...
12      Here's a puppo that seems to be on the fence a...
13      This is Ted. He does his best. Sometimes that'...
14      This is Stuart. He's sporting his favorite fan...
15      This is Oliver. You're witnessing one of his m...
16      This is Jim. He found a fren. Taught him how t...
17      This i

In [14]:
df.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [15]:
#query to see an observation with rating_denominator == 0
df.query('rating_denominator == 0')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [16]:
df.source.sample(10)

1299    <a href="http://vine.co" rel="nofollow">Vine -...
2296    <a href="http://twitter.com/download/iphone" r...
1427    <a href="http://vine.co" rel="nofollow">Vine -...
1911    <a href="http://twitter.com/download/iphone" r...
514     <a href="http://twitter.com/download/iphone" r...
436     <a href="http://twitter.com/download/iphone" r...
1204    <a href="http://twitter.com/download/iphone" r...
1948    <a href="http://twitter.com/download/iphone" r...
2088    <a href="http://twitter.com/download/iphone" r...
1741    <a href="http://twitter.com/download/iphone" r...
Name: source, dtype: object

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

### Quality
#### 1. Remove rows that have been retweeted.
#### 2. Names of the dogs are incorrect (a, an, the and so on can not be the names of the dogs).
#### 3. Text is not fully visible.
#### 4. Remove rows with missing tweet images(expanded_urls).
#### 5. Datatype for timestamp should be date datatype.
#### 6. Some columns of df dataset have wrong datatype as well as missing values. 
#### 7. Doggo, floofer, pupper and puppo have missing values. 
#### 8. Source is not clearly visible as it contains anchor tag, href, etc.
#### 9. Tweet ID 835246439529840640 has a 0 denominator rating.

### Tidiness
#### 1. Dog category variable is in 4 columns.
#### 2. To join all the 3 datasets.

## CLEAN

In [30]:
#to make copy of all dataset and perform the function on them instead of the original dataset
df_clean = df.copy()
df_images_clean = df_images.copy()
df_tweet_json_clean = df_tweet_json.copy()

In [19]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [20]:
df_images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [31]:
df_tweet_json_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2186 entries, 0 to 2185
Data columns (total 3 columns):
tweet_id          2186 non-null object
retweet_count     2186 non-null object
favorite_count    2186 non-null object
dtypes: object(3)
memory usage: 51.3+ KB


## Define

### To join all the 3 datasets

## Code

In [32]:
df_clean.tweet_id = df_clean.tweet_id.astype(str)
df_images_clean.tweet_id = df_images_clean.tweet_id.astype(str)
df_tweet_json_clean.tweet_id = df_tweet_json_clean.tweet_id.astype(str)

df_clean = pd.merge(df_clean, df_images_clean,
                        how = 'inner', on = 'tweet_id')

In [33]:
df_tweet_json_clean = df_tweet_json_clean[~(df_tweet_json_clean.tweet_id.duplicated())]
df_clean=pd.merge(df_clean,df_tweet_json_clean,on='tweet_id',how='left')
df_clean[df_clean.tweet_id=='887473957103951883']
missingcount_id=df_clean.tweet_id[df_clean.retweet_count.isnull()]

## Test

In [34]:
df_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2045  666411507551481857                    NaN                  NaN   
2046  666407126856765440                    NaN                  NaN   
2047  666396247373291520                    NaN                  NaN   
2048  666373753744588802                    NaN                  NaN   
2049  666362758909284353                    NaN                  NaN   
2050  666353288456101888                    NaN                  NaN   
2051  666345417576210432                    NaN                  NaN   
2052  666337882303524864                    NaN                  NaN   
2053  666293911632134144                    NaN                  NaN   
2054  666287406224695296                    NaN                  NaN   
2055  666273097616637952                    NaN                  NaN   
2056  666268910803644416                    NaN                  NaN   
2057  666104133288665088                    NaN                  NaN   
2058  666102155909144576                    NaN                  NaN   
2059  666099513787052032                    NaN                  NaN   
2060  666094000022159362                    NaN                  NaN   
2061  666082916733198337                    NaN                  NaN   
2062  666073100786774016                    NaN                  NaN   
2063  666071193221509120                    NaN                  NaN   
2064  666063827256086533                    NaN                  NaN   
2065  666058600524156928                    NaN                  NaN   
2066  666057090499244032                    NaN                  NaN   
2067  666055525042405380                    NaN                  NaN   
2068  666051853826850816                    NaN         

In [35]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 30 columns):
tweet_id                      2075 non-null object
in_reply_to_status_id         23 non-null float64
in_reply_to_user_id           23 non-null float64
timestamp                     2075 non-null object
source                        2075 non-null object
text                          2075 non-null object
retweeted_status_id           81 non-null float64
retweeted_status_user_id      81 non-null float64
retweeted_status_timestamp    81 non-null object
expanded_urls                 2075 non-null object
rating_numerator              2075 non-null int64
rating_denominator            2075 non-null int64
name                          2075 non-null object
doggo                         2075 non-null object
floofer                       2075 non-null object
pupper                        2075 non-null object
puppo                         2075 non-null object
jpg_url                       2075

## Define

### Combine all the dog columns and put them in one column

## Code

In [36]:
#firstly replace all the data from the doggo, floofer, pupper and puppo columns and replace them wth none
df_clean.doggo=df_clean.doggo.replace('None','')
df_clean.floofer=df_clean.floofer.replace('None','')
df_clean.pupper=df_clean.pupper.replace('None','')
df_clean.puppo=df_clean.puppo.replace('None','')

In [37]:
#then create a new column named dog_category and put them into the new column
df_clean['dog_category']=df_clean.doggo.str.cat(df_clean.floofer).str.cat(df_clean.pupper).str.cat(df_clean.puppo)

In [38]:
#then we drop all four previous columns as ther are of no use now
df_clean=df_clean.drop(['doggo','floofer','pupper','puppo'],axis=1)
df_clean.dog_category=df_clean.dog_category.replace('',np.nan)

## Test

In [39]:
df_clean.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1202  697990423684476929                    NaN                  NaN   
1043  711363825979756544                    NaN                  NaN   
886   739544079319588864                    NaN                  NaN   
1854  670086499208155136                    NaN                  NaN   
1115  705591895322394625                    NaN                  NaN   
1038  711968124745228288                    NaN                  NaN   
1484  679148763231985668                    NaN                  NaN   
1059  709852847387627521                    NaN                  NaN   
1532  677228873407442944                    NaN                  NaN   
255   834209720923721728                    NaN                  NaN   

                      timestamp  \
1202  2016-02-12 03:47:39 +0000   
1043  2016-03-20 01:28:47 +0000   
886   2016-06-05 19:47:03 +0000   
1854  2015-11-27 03:47:25 +0000   
1115  2016-03-04 03:13:11 +0000   
1038  2016-03-21 17:30:03 +0000   
1484  2015-12-22 03:57:37 +0000   
1059  2016-03-15 21:24:41 +0000   
1532  2015-12-16 20:48:40 +0000   
255   2017-02-22 01:14:30 +0000   

                                                 source  \
1202  <a href="http://twitter.com/download/iphone" r...   
1043  <a href="http://twitter.com/download/iphone" r...   
886   <a href="http://twitter.com/download/iphone" r...   
1854  <a href="http://twitter.com/download/iphone" r...   
1115  <a href="http://twitter.com/download/iphone" r...   
1038  <a href="http://twitter.com/download/iphone" r...   
1484  <a href="http://twitter.com/download/iphone" r...   
1059  <a href="http://twitter.com/download/iphone" r...   
1532  <a href="http://twitter.com/download/iphone" r...   
255   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1202  This is Doug. He's a Draconian Jabbawockee. Ra...                  NaN   
1043  "Please, no puparazzi" 11/10 https://t.co/nJIX...                  NaN   
886   This... is a Tyrannosaurus rex. We only rate d...                  NaN   
1854  "Hi yes this is dog. I can't help with that s-...                  NaN   
1115  "Ma'am, for the last time, I'm not authorized ...                  NaN   
1038  Meet Winston. He's trapped in a cup of coffee....                  NaN   
1484  I know everyone's excited for Christmas but th...                  NaN   
1059  *lets out a tiny whimper and then collapses* ....                  NaN   
1532  Say hello to Chuq. He just wants to fit in. 11...                  NaN   
255   This is Wilson. He's aware that he has somethi...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1202                       NaN                        NaN   
1043                       NaN                        NaN   
886                        NaN                        NaN   
1854                       NaN                        NaN   
1115                       NaN                        NaN   
1038                       NaN                        NaN   
1484                       NaN                        NaN   
1059                       NaN                        NaN   
1532                       NaN                        NaN   
255                        NaN                        NaN   

                                          expanded_urls      ...       p1_dog  \
1202  https://twitter.com/dog_rates/status/697990423...      ...         True   
1043  https://twitter.com/dog_rates/status/711363825...      ...         True   
886   https://twitter.com/dog_rates/status/739544079...      ...         True   
1854  https://twitter.com/dog_rates/status/670086499...      ...         True   
1115  https://twitter.com/dog_rates/status/705591895...      ...         True   
1038  https://twitter.com/dog_rates/status/711968124...      ...        False   
1484  https://twitter.com/dog_rates/status/679148763...      ...   

## Define

### To remove the retweeted observations we also have to remove retweeted_satus_id, retweeted_satus_user_id and retweeted_satus_timestamp columns as they will no longer be needed

## Code

In [40]:
df_clean = df_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],axis=1)

## Test

In [41]:
df_clean.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1491  678800283649069056                    NaN                  NaN   
1987  667524857454854144                    NaN                  NaN   
734   758740312047005698                    NaN                  NaN   
115   865718153858494464                    NaN                  NaN   
628   775898661951791106                    NaN                  NaN   
332   822647212903690241                    NaN                  NaN   
1005  716080869887381504                    NaN                  NaN   
370   817415592588222464                    NaN                  NaN   
1358  685641971164143616                    NaN                  NaN   
1004  716285507865542656                    NaN                  NaN   

                      timestamp  \
1491  2015-12-21 04:52:53 +0000   
1987  2015-11-20 02:08:22 +0000   
734   2016-07-28 19:06:01 +0000   
115   2017-05-19 23:57:46 +0000   
628   2016-09-14 03:27:11 +0000   
332   2017-01-21 03:29:14 +0000   
1005  2016-04-02 01:52:38 +0000   
370   2017-01-06 17:00:38 +0000   
1358  2016-01-09 01:59:19 +0000   
1004  2016-04-02 15:25:47 +0000   

                                                 source  \
1491  <a href="http://twitter.com/download/iphone" r...   
1987  <a href="http://twitter.com" rel="nofollow">Tw...   
734   <a href="http://twitter.com/download/iphone" r...   
115   <a href="http://twitter.com/download/iphone" r...   
628   <a href="http://twitter.com/download/iphone" r...   
332   <a href="http://twitter.com/download/iphone" r...   
1005  <a href="http://twitter.com/download/iphone" r...   
370   <a href="http://twitter.com/download/iphone" r...   
1358  <a href="http://twitter.com/download/iphone" r...   
1004  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
1491  Here's a pupper with some mean tan lines. Snaz...   
1987  Another topnotch dog. His name is Big Jumpy Ra...   
734   Meet Boston. He's worried because his tongue w...   
115   Meet Boomer. He's just checking pup on you. Ho...   
628   RT @dog_rates: Like father (doggo), like son (...   
332   RT @dog_rates: This is Paisley. She really wan...   
1005  Here's a super majestic doggo and a sunset 11/...   
370   Meet Strudel. He's rather h*ckin pupset that y...   
1358  This is Otis. He just passed a cop while going...   
1004  This is Stephanus. She stays woke. 12/10 https...   

                                          expanded_urls  rating_numerator  \
1491  https://twitter.com/dog_rates/status/678800283...                12   
1987  https://twitter.com/dog_rates/status/667524857...                12   
734   https://twitter.com/dog_rates/status/758740312...                12   
115   https://twitter.com/dog_rates/status/865718153...                13   
628   https://twitter.com/dog_rates/status/733109485...                12   
332   https://twitter.com/dog_rates/status/822489057...                13   
1005  https://twitter.com/dog_rates/status/716080869...                11   
370   https://www.gofundme.com/help-strudel-walk-aga...                11   
1358  https://twitter.com/dog_rates/status/685641971...                 7   
1004  https://twitter.com/dog_rates/status/716285507...                12   

      rating_denominator       name      ...      p1_dog  \
1491                  10       None      ...        True   
1987                  10       None      ...       False   
734                   10     Boston      ...        True   
115                   10     Boomer      ...        True   
628                   10       None      ...        True   
332                   10    Paisley      ...        True   
1005                  10       None      ...        True   
370                   10    Strudel      ...        True   
1358                  10       Otis      ...        True   
1004                  10  Stephanus      ...        True   

                           p

In [42]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 24 columns):
tweet_id                 2075 non-null object
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2075 non-null object
source                   2075 non-null object
text                     2075 non-null object
expanded_urls            2075 non-null object
rating_numerator         2075 non-null int64
rating_denominator       2075 non-null int64
name                     2075 non-null object
jpg_url                  2075 non-null object
img_num                  2075 non-null int64
p1                       2075 non-null object
p1_conf                  2075 non-null float64
p1_dog                   2075 non-null bool
p2                       2075 non-null object
p2_conf                  2075 non-null float64
p2_dog                   2075 non-null bool
p3                       2075 non-null object
p3_conf                  2

## Define

### Remove rows with missing tweet images(expanded_urls).

## Code

In [43]:
# query expanded_urls that are not null
df_clean = df_clean[df_clean.expanded_urls.notnull()]

## Test

In [44]:
sum(df_clean.expanded_urls.isnull())

0

In [45]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 24 columns):
tweet_id                 2075 non-null object
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2075 non-null object
source                   2075 non-null object
text                     2075 non-null object
expanded_urls            2075 non-null object
rating_numerator         2075 non-null int64
rating_denominator       2075 non-null int64
name                     2075 non-null object
jpg_url                  2075 non-null object
img_num                  2075 non-null int64
p1                       2075 non-null object
p1_conf                  2075 non-null float64
p1_dog                   2075 non-null bool
p2                       2075 non-null object
p2_conf                  2075 non-null float64
p2_dog                   2075 non-null bool
p3                       2075 non-null object
p3_conf                  2

## Define

### Text not fully visible

## Code

In [46]:
#text column is widened to full length
pd.set_option('display.max_colwidth', -1)

## Test

In [47]:
df_clean.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
891   738537504001953792 NaN                    NaN                    
212   842765311967449089 NaN                    NaN                    
728   759197388317847553 NaN                    NaN                    
236   837471256429613056 NaN                    NaN                    
1400  683391852557561860 NaN                    NaN                    
1915  668892474547511297 NaN                    NaN                    
720   760252756032651264 NaN                    NaN                    
545   789268448748703744 NaN                    NaN                    
921   732726085725589504 NaN                    NaN                    
1921  668655139528511488 NaN                    NaN                    

                      timestamp  \
891   2016-06-03 01:07:16 +0000   
212   2017-03-17 15:51:22 +0000   
728   2016-07-30 01:22:17 +0000   
236   2017-03-03 01:14:41 +0000   
1400  2016-01-02 20:58:09 +0000   
1915  2015-11-23 20:42:48 +0000   
720   2016-08-01 23:15:56 +0000   
545   2016-10-21 00:53:56 +0000   
921   2016-05-18 00:14:46 +0000   
1921  2015-11-23 04:59:42 +0000   

                                                                                  source  \
891   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
212   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
728   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
236   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1400  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1915  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
720   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
545   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
921   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1921  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                           text  \
891   This is Bayley. She fell asleep trying to escape her evil fence enclosure. 11/10 night night puppo https://t.co/AxSiqAKEKu                                                  
212   Meet Indie. She's not a fan of baths but she's definitely a fan of hide &amp; seek. 12/10 click the link to help Indie\n\nhttps://t.co/fvGkIuAlFK https://t.co/kiCFtmJd7l   
728   This is Luna. She's just heckin precious af I have nothing else to say. 12/10 https://t.co/gQH2mmKIJW                                                                       
236   This is Vincent. He's suave as h*ck. Will be your copilot this evening. Claims he doesn't need to look at the directions. 12/10 https://t.co/u51tzXSVi3                     
1400  Say hello to Jack (pronounced "Kevin"). He's a Virgo Episcopalian. Can summon rainbows. 11/10 magical as hell https://t.co/YHXrdUTHd6                                       
1915  This is Ruffles. He is an Albanian Shoop Da Whoop. He just noticed the camera. Patriotic af. Classy hardwood. 11/10 https://t.co/HyDpTU5Jhj                                 
720   This is Fido. He can tell the weather. Not good at fetch tho. Never comes when called. 4/10 would probably still pet https://t.co/4gOv2Q3iKP                                
545   This is Stella. She's happier than I will ever be. 10/10 would trade lives with https://t.co/JSs2bfDtTS                                                                     
921   This is Aldrick. He looks wise af. Also exceptionally fluffy. Only two legs tho (unfortunate). 11/10 would still hug https://t.co/QwiCVLPMNL                                
1921  Say hello to Winifred. He is

## Define

### To correct the names of the dogs by giving them either correct names or giving them None value.

## Code

In [48]:
df_clean['name'].replace("the","None", inplace=True)
df_clean['name'].replace("light","Lightning", inplace=True)
df_clean['name'].replace("life","None", inplace=True)
df_clean['name'].replace("an","None", inplace=True)
df_clean['name'].replace("a","None", inplace=True)
df_clean['name'].replace("by","None", inplace=True)
df_clean['name'].replace("actually","None", inplace=True)
df_clean['name'].replace("just","None", inplace=True)
df_clean['name'].replace("getting","None", inplace=True) 
df_clean['name'].replace("infuriating","None", inplace=True) 
df_clean['name'].replace("old","None", inplace=True) 
df_clean['name'].replace("all","None", inplace=True) 
df_clean['name'].replace("this","None", inplace=True) 
df_clean['name'].replace("very","None", inplace=True) 
df_clean['name'].replace("mad","None", inplace=True) 
df_clean['name'].replace("not","None", inplace=True)
df_clean['name'].replace("one","None", inplace=True)
df_clean['name'].replace("my","None", inplace=True)
df_clean['name'].replace("O","olly", inplace=True)
df_clean['name'].replace("quite","None", inplace=True)
df_clean['name'].replace("such","None", inplace=True)

## Test

In [49]:
df_clean.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
407   812781120811126785 NaN                    NaN                    
1528  677328882937298944 NaN                    NaN                    
1124  704859558691414016 NaN                    NaN                    
1458  680191257256136705 NaN                    NaN                    
1751  671874878652489728 NaN                    NaN                    
1884  669393256313184256 NaN                    NaN                    
364   818259473185828864 NaN                    NaN                    
791   750429297815552001 NaN                    NaN                    
1606  675109292475830276 NaN                    NaN                    
1418  682259524040966145 NaN                    NaN                    

                      timestamp  \
407   2016-12-24 22:04:54 +0000   
1528  2015-12-17 03:26:04 +0000   
1124  2016-03-02 02:43:09 +0000   
1458  2015-12-25 01:00:07 +0000   
1751  2015-12-02 02:13:48 +0000   
1884  2015-11-25 05:52:43 +0000   
364   2017-01-09 00:53:55 +0000   
791   2016-07-05 20:41:01 +0000   
1606  2015-12-11 00:26:12 +0000   
1418  2015-12-30 17:58:40 +0000   

                                                                                  source  \
407   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1528  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1124  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1458  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1751  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1884  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
364   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
791   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1606  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1418  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                           text  \
407   This is Swagger. He's the Cleveland Browns ambassador. Hype as h*ck after that first win today. 10/10 https://t.co/lXFM1l22bG               
1528  This is Derek. All the dogs adore Derek. He's a great guy. 10/10 really solid pup https://t.co/KgcsGNb61s                                   
1124  Here is a heartbreaking scene of an incredible pupper being laid to rest. 10/10 RIP pupper https://t.co/81mvJ0rGRu                          
1458  Here's a sleepy Christmas pupper 11/10 https://t.co/KXg0f8GNQ9                                                                              
1751  This is Mia. She was specifically told not get on top of the hutch or play in the fridge. 10/10 what a rebel https://t.co/3J7wkwW4FG        
1884  Meet Ronduh. She's a Finnish Checkered Blitzkrieg. Ears look fake. Shoes on point. 10/10 would pet extra well https://t.co/juktj5qiaD       
364   This is Florence. He saw the same snap you sent him on your story. Pretty pupset with you. 12/10 https://t.co/rnkvT2kvib                    
791   This is Arnie. He's a Nova Scotian Fridge Floof. Rare af. 12/10 https://t.co/lprdOylVpS                                                     
1606  C'mon guys. We've been over this. We only rate dogs. This is a cow. Please only submit dogs. Thank you...... 9/10 https://t.co/WjcELNEqN2   
1418  Meet Jax. He's an Iglesias Hufflepoof. Quite the jokester. Takes it too far sometimes. Can be very hurtful. 9/10 https://t.co/i5TeG0KYcW    

                                                                                                                                                                                        expanded_urls  \
407  

## Define

### Source is not clearly visible as it contains anchor tag, href, etc.

## Code

In [50]:
# extract the source that is between the anchor tags
df_clean.source = df_clean.source.str.replace('<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>','Twitter for iPhone')
df_clean.source = df_clean.source.str.replace('<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>','Vine - make a Scene')
df_clean.source = df_clean.source.str.replace('<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>','Twitter Web Client')
df_clean.source = df_clean.source.str.replace('<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>','Tweet Deck')

## Test

In [51]:
df_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
19    888202515573088257 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
...                  ...  ..                     ..                    
2045  666411507551481857 NaN                    NaN                    
2046  666407126856765440 NaN                    NaN                    
2047  666396247373291520 NaN                    NaN                    
2048  666373753744588802 NaN                    NaN                    
2049  666362758909284353 NaN                    NaN                    
2050  666353288456101888 NaN                    NaN                    
2051  666345417576210432 NaN                    NaN                    
2052  666337882303524864 NaN                    NaN                    
2053  666293911632134144 NaN                    NaN                    
2054  666287406224695296 NaN                    NaN                    
2055  666273097616637952 NaN                    NaN                    
2056  666268910803644416 NaN                    NaN                    
2057  666104133288665088 NaN                    NaN                    
2058  666102155909144576 NaN                    NaN                    
2059  666099513787052032 NaN                    NaN                    
2060  666094000022159362 NaN                    NaN                    
2061  666082916733198337 NaN                    NaN                    
2062  666073100786774016 NaN                    NaN                    
2063  666071193221509120 NaN                    NaN                    
2064  666063827256086533 NaN                    NaN                    
2065  666058600524156928 NaN                    NaN                    
2066  666057090499244032 NaN                    NaN                    
2067  666055525042405380 NaN                    NaN                    
2068  666051853826850816 NaN                    NaN     

## Define

### Tweet ID 835246439529840640 has a 0 denominator rating.

## Code

In [52]:
#remove observation who has rating_denomibator == 0
df_clean = df_clean.query('rating_denominator != 0')

## Test

In [53]:
df_clean.query('rating_denominator == 0')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog, retweet_count, favorite_count, dog_category]
Index: []

[0 rows x 24 columns]

In [54]:
df_clean.rating_denominator.value_counts()

10     2057
50     3   
11     2   
80     2   
7      1   
170    1   
150    1   
130    1   
120    1   
110    1   
90     1   
70     1   
40     1   
20     1   
2      1   
Name: rating_denominator, dtype: int64

## Define

### Datatype for timestamp should be date datatype.

## Code

In [55]:
df_clean['timestamp'] = pd.to_datetime(df_clean['timestamp'])

## Test

In [56]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 24 columns):
tweet_id                 2075 non-null object
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                2075 non-null datetime64[ns]
source                   2075 non-null object
text                     2075 non-null object
expanded_urls            2075 non-null object
rating_numerator         2075 non-null int64
rating_denominator       2075 non-null int64
name                     2075 non-null object
jpg_url                  2075 non-null object
img_num                  2075 non-null int64
p1                       2075 non-null object
p1_conf                  2075 non-null float64
p1_dog                   2075 non-null bool
p2                       2075 non-null object
p2_conf                  2075 non-null float64
p2_dog                   2075 non-null bool
p3                       2075 non-null object
p3_conf           

## Define

### Some columns of df dataset have wrong datatype as well as missing values.

## Code

In [57]:
df_clean['in_reply_to_status_id'] = df_clean['in_reply_to_status_id'].astype('str')
df_clean['in_reply_to_user_id'] = df_clean['in_reply_to_user_id'].astype('str')

## Test

In [58]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 24 columns):
tweet_id                 2075 non-null object
in_reply_to_status_id    2075 non-null object
in_reply_to_user_id      2075 non-null object
timestamp                2075 non-null datetime64[ns]
source                   2075 non-null object
text                     2075 non-null object
expanded_urls            2075 non-null object
rating_numerator         2075 non-null int64
rating_denominator       2075 non-null int64
name                     2075 non-null object
jpg_url                  2075 non-null object
img_num                  2075 non-null int64
p1                       2075 non-null object
p1_conf                  2075 non-null float64
p1_dog                   2075 non-null bool
p2                       2075 non-null object
p2_conf                  2075 non-null float64
p2_dog                   2075 non-null bool
p3                       2075 non-null object
p3_conf         

## STORING DATA

In [59]:
df_clean.to_csv('twitter_archive_master.csv', index=False, encoding = 'utf-8')

In [60]:
we_rate_dogs = pd.read_csv('twitter_archive_master.csv')

In [62]:
we_rate_dogs

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
19    888202515573088257 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
...                  ...  ..                     ..                    
2045  666411507551481857 NaN                    NaN                    
2046  666407126856765440 NaN                    NaN                    
2047  666396247373291520 NaN                    NaN                    
2048  666373753744588802 NaN                    NaN                    
2049  666362758909284353 NaN                    NaN                    
2050  666353288456101888 NaN                    NaN                    
2051  666345417576210432 NaN                    NaN                    
2052  666337882303524864 NaN                    NaN                    
2053  666293911632134144 NaN                    NaN                    
2054  666287406224695296 NaN                    NaN                    
2055  666273097616637952 NaN                    NaN                    
2056  666268910803644416 NaN                    NaN                    
2057  666104133288665088 NaN                    NaN                    
2058  666102155909144576 NaN                    NaN                    
2059  666099513787052032 NaN                    NaN                    
2060  666094000022159362 NaN                    NaN                    
2061  666082916733198337 NaN                    NaN                    
2062  666073100786774016 NaN                    NaN                    
2063  666071193221509120 NaN                    NaN                    
2064  666063827256086533 NaN                    NaN                    
2065  666058600524156928 NaN                    NaN                    
2066  666057090499244032 NaN                    NaN                    
2067  666055525042405380 NaN                    NaN                    
2068  666051853826850816 NaN                    NaN     